ndless-sdk/bin/nspire-tools

```diff
@@ -2,41 +2,61 @@
 
 function usage() {
 	echo "Usage: nspire-tools <command> <args>"
-	echo -e "   new <program>:   Generate a standard Makefile for <program>.tns"
-	echo -e "   main:            Create an empty main.c"
+	echo "   new <program>:   Generate a standard Makefile for <program>.tns"
+	echo "   main:            Create an empty main.c"
+	echo "   path:            Print out the path of the ndless installation directory"
+	echo "   gcc <args...>:   Run arm-none-eabi-gcc with arguments. Usually prefer the nspire-gcc command over this."
+	echo "   toolchainbin:    Print out the path of the directory that has the binaries for Ndless' arm-none-eabi toolchain."
 }
 
 
 UNAME="$(uname -s | tr A-Z a-z)"
-DIRNAME="$(dirname "$0")"
 
 # DIRNAME may not be in the toolchain directory if this script was symlinked.
 # try to find the symlink target	
-which readlink >/dev/null 2>&1 && readlink "$0" >/dev/null 2>&1 && {
-    SYMLINK="$(readlink "$0")"
-    
-    # On MacOS X, the readlink command hasn't the same behavior than the others operating systems based on Unix.
-    if [ "$UNAME" == "darwin" ]; then
-    	DIRNAME="$(dirname "$SYMLINK")"
-    else
-    	DIRNAME="${SYMLINK}/.."
-    fi
-}
+{
+  if which realpath; then
+    alias realpath=realpath
+  elif readlink -f /; then
+    alias realpath='readlink -f'
+  elif which readlink; then
+    alias realpath=readlink
+  else
+    alias realpath=false
+  fi
+
+  real0="$(realpath "$0")" || real0=$0
+  DIRNAME="$(dirname "$real0")"
+} >/dev/null 2>&1
+
 
 # Cygwin owns a dedicated command to find the symlink target
 if [ "$UNAME" = "cygwin" ]; then
-	DIRNAME="$(cygpath --path --windows "$DIRNAME")"
+	DIRNAME="$(cygpath --path --windows "$BINPATH")"
 fi
+NDLESS="$(dirname "$DIRNAME")"
 
 if [ $# -eq 0 ]; then
 	usage
 	exit 0
 fi
 
-case "$1" in
+# following functions can be patched when packaging/distributing
+toolchainbin() {
+	echo "$NDLESS/toolchain/install/bin"
```

Comment:
An idea: Make this print an environment variable instead if set, so that it can be overwritten simply by doing an `export`.

Patching should ideally not be necessary.

Reply:
I think patching is kind of unavoidable in some cases; requiring a user to set an environment variable after they install a package isn't very ergonomic, and different distros have different ways of organizing packages. Plus, a patch for these functions is pretty minimal, my aur patch for this section is just:

```patch
 # following functions can be patched when packaging/distributing
 toolchainbin() {
-	echo "$NDLESS/toolchain/install/bin"
+	echo "$NDLESS/bin"
 }
 zehn_loader_path() {
-	MAKEFLAGS='' make -C "${NDLESS}/tools/zehn_loader" -s all
 	local suffix=$1
-	echo "$NDLESS/tools/zehn_loader/zehn_loader$suffix.tns"
+	echo "$NDLESS/lib/zehn_loader$suffix.tns"
 }
```

Reply:
Whoops, accidentially marked as resolved...

I think that a patch is avoidable in this case. You can do that in a variety of ways, like adding a file into `/etc/profile.d/` or placing a wrapper into `/usr/bin/` which just sets the variable.

It should work just fine to use a package provided toolchain with the latest SDK checkout from git.

Reply:
What about with `zehn_loader_path`? It still needs to be able to pass whether it's compressed to the function; how would that work with an environment variable?

Reply:
If `zehn_loader_path` only returns the path to the precompiled binaries, that should work fine.

Reply:
I did basically that, `$_NDLESS_ZEHN_PATH` has the `zehn_loader[_compressed].tns` files.